### PR TITLE
Refine the API in the public header

### PR DESCRIPTION
**Pull request by ChinYikMing**
The following should be included in an emulator's simple and clear public API:
1. create/init core
2. run emulation
3. delete/destroy core

Other components, including as memory, file systems, program data, etc., should be abstracted from the user, as a result, setting a configuration value (`vm_attr_t`) is sufficient. The user should manage about memory (`state_t`) and elf stuff before this PR. The user may just construct a core, run it, and shut it down after this PR, so they won't need to worry about them anymore.

For stdio remapping, `rv_remap_stdstream` function is introduced.

The `vm_attr_t` has multiple fields and they are commented clearly in the code.

elf is reopened in `dump_test_signature` because elf is allocated during `rv_create`. It is acceptable to reopen elf since it is only for testing. Print inferior exit code to console inside main instead of `syscall_exit` because the actual usage of exit code depends on applications of using riscv public API.

The `io` interface is not changed in this PR because it could maybe reused with `semu` in some way, still need to be investigated. Also, Logging feature and system emulator integration are not implemented yet.

### CODE REVIEW

File: src/syscall_sdl.c
Comment (ChinYikMing) [inline src/syscall_sdl.c]: `s` should be renamed to `vm_attr`

File: src/syscall_sdl.c
Comment (ChinYikMing) [inline src/syscall_sdl.c]: `s` should be renamed to `vm_attr`

File: src/syscall_sdl.c
Comment (ChinYikMing) [inline src/syscall_sdl.c]: `s` should be renamed to `vm_attr`

File: src/syscall_sdl.c
Comment (ChinYikMing) [inline src/syscall_sdl.c]: `s` should be renamed to `vm_attr`

File: src/syscall_sdl.c
Comment (ChinYikMing) [inline src/syscall_sdl.c]: `s` should be renamed to `vm_attr`

File: src/emulate.c
Comment (jserv) [inline src/emulate.c]: We can introduce a helper macro similar to how `PRIV` is used in semu. That is,
```c
#define PRIV(x) ((vm_attr_t *) x->userdata)
```

File: src/main.c
Comment (jserv) [inline src/main.c]: Name the instance of `vm_attr_t` simply as `attr`.

File: src/riscv.c
Comment (jserv) [inline src/riscv.c]: Quote [Wikipedia](https://en.wikipedia.org/wiki/Standard_streams):
> In computer programming, standard streams are preconnected input and output communication channels between a computer program and its environment when it begins execution. The three input/output (I/O) connections are called standard input (stdin), standard output (stdout) and standard error (stderr).

In the proposed changed code, the use of stdio is confusing.

File: src/syscall_sdl.c
Comment (jserv) [reply src/syscall_sdl.c]: Or, `attr`.

**Comment by ChinYikMing** [comment]
As discuss before, exit code shall be stored and maybe not to output to console directly. 

Unfortunately, CI/host-x64 and CI/host-arm run `make check` and `make misalign` for testing, and the compare signature includes "inferior exit code 0" during exit but I do not output them in this PR. Shall we modify `make check` and `make misalign` ?

File: src/syscall.c
Comment (jserv) [inline src/syscall.c]: The line `vm_attr_t *vm_attr = rv_userdata(rv);` might be perplexing due to the similar naming of the object type and the `vm_attr`, while the value comes from the `rv_userdata` call. This suggests a presumed direct link between `rv_userdata` and `vm_attr_t`. However, this connection is not explicitly clarified in the API documentation, leading to potential confusion regarding their relationship.

**Comment by jserv** [comment]
> As discuss before, exit code shall be stored and maybe not to output to console directly.
> Unfortunately, CI/host-x64 and CI/host-arm run `make check` and `make misalign` for testing, and the compare signature includes "inferior exit code 0" during exit but I do not output them in this PR. Shall we modify `make check` and `make misalign` ?

After a RISC-V program ends, it is important to verify the exit code (i.e., `$?`). Consequently, modifications to 'check' and related build targets are necessary to accommodate this.

**Comment by ChinYikMing** [comment]
> After a RISC-V program ends, it is important to verify the exit code (i.e., `$?`). Consequently, modifications to 'check' and related build targets are necessary to accommodate this.

Sure, a build macro can be added to `syscall_exit` and print  the exit code only on test for passing CI.

**Comment by visitorckw** [comment]
Due to the memory size being changed to a runtime-defined value, do we need to check if the addresses of load and store instructions exceed the specified range?

File: src/riscv.c
Comment (ChinYikMing) [reply src/riscv.c]: I think `rv_remap_stream(riscv_t *rv, int fd, FILE *fptr)` is more comprehensive. More comments can be added to classify the parameters.

Comment (jserv) [reply src/riscv.c]: How about `rv_remap_stdstream(riscv_t *rv, int fd, FILE *filev)`?
Here `filev` means a vector of `FILE *`.

Comment (ChinYikMing) [reply src/riscv.c]: A fd is given in a call, is it necessary required a vector of `FILE *`?

Comment (jserv) [reply src/riscv.c]: > A fd is given in a call, is it necessary required a vector of `FILE *`?

No, my intention was to encapsulate the pair of a file descriptor and standard streams. This was meant to form a vector, but I misplaced the parameters.

Comment (ChinYikMing) [reply src/riscv.c]: > How about rv_remap_stdstream(riscv_t *rv, int fd, FILE *filev)?

Much more sense than `rv_register_stdio`.

> > A fd is given in a call, is it necessary required a vector of `FILE *`?
> 
> No, my intention was to encapsulate the pair of a file descriptor and standard streams. This was meant to form a vector, but I misplaced the parameters.

Got it. So, the signature becomes `rv_remap_stdstream(riscv_t *rv, int *fdv, FILE *filev)`. We could assume the length of the vector is 3, otherwise undefined behavior. If any `fd` in `fdv` is -1, it fallback to use default stdio.

Comment (jserv) [reply src/riscv.c]: > So, the signature becomes `rv_remap_stdstream(riscv_t *rv, int *fdv, FILE *filev)`. We could assume the length of the vector is 3, otherwise undefined behavior. If any `fd` in `fdv` is -1, it fallback to use default stdio.

To mitigate the risk of passing null or malformed pointers, we could introduce a structure, `fd_stream_pair`, to encapsulate file descriptors and `FILE` pointers. This structure can then be reliably passed as a pointer to the `rv_remap_stdstream` function, ensuring safer and more organized handling of these elements.

**Comment by jserv** [comment]
I would like to express a concern regarding the `rv_userdata` function being part of the public API. Originally, it was designed for internal implementation, which suggests there is no necessity to make `rv_userdata` accessible to users interacting with the public API. It might be more appropriate to keep this function internal to maintain clarity and simplicity in the API's usage.

File: src/syscall.c
Comment (ChinYikMing) [reply src/syscall.c]: Added more comments about them. Hopefully it is clearer now.

**Comment by ChinYikMing** [comment]
> Sure, a build macro can be added to `syscall_exit` and print the exit code only on test for passing CI.

The PRINT_EXIT_CODE build macro do not used in default build, thus make CI/host-x64 or CI/host-arm64 failed. Shall I add it to  default build?  Personally, I am prefer default build without PRINT_EXIT_CODE.

Another way to pass CI is that modifying the CI flow, build `rv32emu` with PRINT_EXIT_CODE  in CI tests.

**Comment by jserv** [comment]
> > Sure, a build macro can be added to `syscall_exit` and print the exit code only on test for passing CI.
> 
> The PRINT_EXIT_CODE build macro do not used in default build, thus make CI/host-x64 or CI/host-arm64 failed. Shall I add it to default build? Personally, I am prefer default build without PRINT_EXIT_CODE.
> Another way to pass CI is that modifying the CI flow, build `rv32emu` with PRINT_EXIT_CODE in CI tests.

Drop macro `PRINT_EXIT_CODE` because it is not necessary to have multiple builds. Instead, let the shell or builtin commands to check exit code of RISC-V program emulated.

**Comment by ChinYikMing** [comment]
> Due to the memory size being changed to a runtime-defined value, do we need to check if the addresses of load and store instructions exceed the specified range?

Sound reasonable. Maybe in future PR.

File: Makefile
Comment (ChinYikMing) [inline Makefile]: May I know the original input of this hash value?

Comment (ChinYikMing) [reply Makefile]: Maybe I guessed it

**Comment by ChinYikMing** [comment]
> Drop macro `PRINT_EXIT_CODE` because it is not necessary to have multiple builds. Instead, let the shell or builtin commands to check exit code of RISC-V program emulated.

For minimal changes, printing `inferior exit code %d` inside `main` is able to pass the CI tests.

**Comment by ChinYikMing** [comment]
> I would like to express a concern regarding the `rv_userdata` function being part of the public API. Originally, it was designed for internal implementation, which suggests there is no necessity to make `rv_userdata` accessible to users interacting with the public API. It might be more appropriate to keep this function internal to maintain clarity and simplicity in the API's usage.

Actually, we could introduce some helper functions or getter functions to access the specific field of userdata without losing the encapsulation. For example, some helper functions for memory viewing instead of returning whole userdata.

File: src/main.c
Comment (RinHizakura) [inline src/main.c:226]: Since we are going to make these parameters user-defined now, I will suggest having some checks (`check_vm_attr()`) on the `vm_attr_t` before the emulator creation stage, otherwise bad parameters can lead to a crash which is not easy to figure out quickly. For example:  `mem_size` should be larger than `stack_size + args_offset_size`.

Note that `mem_size` should also be larger than the ELF mapping range to load correctly, but this may not be easy to check before the emulator creation stage. As an alternative, maybe applying the range check for `memory_write` at the emulator running stage is a simple idea to solve this. We may need to take care of the possible performance drop for the extra comparison on this solution.

File: src/main.c
Comment (RinHizakura) [inline src/main.c]: Why don't we just realize the elf name by `PRIV(rv)`? It seems that we don't need to pass `opt_prog_name` again here.

File: src/riscv.c
Comment (RinHizakura) [inline src/riscv.c]: To make the implementation consistent, replace with `assert(rv)`.

File: src/riscv.c
Comment (RinHizakura) [inline src/riscv.c]: Similarly, replace with `assert(attr)`.

File: src/riscv.c
Comment (RinHizakura) [inline src/riscv.c]: Let's `assert(attr)` here since we'll access it without checking in the future.

File: src/riscv.c
Comment (RinHizakura) [inline src/riscv.c]: Maybe it's not a great idea to use 0, 1, and 2 here directly. @jserv Do you think we should define macros for each and expand them to the correct value considering the operating system(e.g.  `STDIN_FILENO` for POSIX and _fileno(stdin) on WIN)?

Comment (jserv) [reply src/riscv.c]: > Do you think we should define macros for each and expand them to the correct value considering the operating system(e.g. `STDIN_FILENO` for POSIX and _fileno(stdin) on WIN)?

Agree. We shall proceed in a portable way.

File: src/main.c
Comment (ChinYikMing) [reply src/main.c]: Nice catch!

File: src/riscv.c
Comment (ChinYikMing) [reply src/riscv.c]: See [commit](https://github.com/sysprog21/rv32emu/pull/340/commits/f953ac1e051558623ae3d1bfd4cee70077e47c53#diff-f99b8d1f48ef2257a9bbbe6b487c621f74ef233c00858df5a37b06e89441e836)

File: src/main.c
Comment (ChinYikMing) [reply src/main.c]: As `PRIV(rv)` requires internal structure, `rv_userdata(rv)` as alternative.

File: src/riscv.c
Comment (jserv) [inline src/riscv.c]: Refine the use of `stdio_fd`.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: The macro `PRIV` should not appear in public header, `riscv.h`.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: To maintain a consistent API naming scheme and to prevent potential symbol conflicts, it is advisable to choose a name other than `run` for the function.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: In this context, `rv` represents the instance of the RISC-V emulator.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: Let's hold off on including the system emulation components until we are fully prepared to merge them.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: Ditto. Get rid of `vm_sys` and place proper TODO comments.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: Unify the naming in the form of `fd_stdxxx`. That is,
```c
int fd_stdin, fd_stdout, fd_stderr;
```

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: Refine the comment.

**Comment by ChinYikMing** [comment]
> Actually, we could introduce some helper functions or getter functions to access the specific field of userdata without losing the encapsulation. For example, some helper functions for memory viewing instead of returning whole userdata.

Shall we drop it for now? Later on, we can think about what fields are gonna export to applications.

Note that the `run_and_trace` function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop `rv_userdata`, the `run_and_trace` can only get program name by passing `opt_prog_name` or `attr` to it.

File: src/riscv.h
Comment (ChinYikMing) [reply src/riscv.h]: Stored it in `riscv_private.h`

File: src/riscv.h
Comment (ChinYikMing) [reply src/riscv.h]: How about `rv_start_emu`?

Comment (jserv) [reply src/riscv.h]: > How about `rv_start_emu`?

Opt for concise and meaningful names. e.g., `rv_mainloop` and `rv_run`.

**Comment by jserv** [comment]
> Note that the run_and_trace function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop rv_userdata, the run_and_trace can only get program name by passing opt_prog_name or attr to it.

We could streamline the `run` and `run_and_trace` functions by treating `run_and_trace` as a specialized variant of `rv_run`, activated through specific callback functions or flags. This strategy would enable us to eliminate the need for exposing `rv_userdata` directly.

File: src/riscv.h
Comment (jserv) [reply src/riscv.h]: Check the manner of GLib. The [GMainLoop struct](https://docs.gtk.org/glib/struct.MainLoop.html) is an opaque data type representing the main event loop of a GLib or GTK application.

File: src/main.c
Comment (ChinYikMing) [reply src/main.c:226]: Nice perspective. I think validation of `vm_attr_t` could be complex. For example, shall we check the invalid file during validation or propagate to `elf_open` to verify the invalidation. Similar situation for memory, if the user-provided memory size is not enough to run emulation, shall we simply abort before `rv_create` or propagate to segmentation fault. We may need more discussion about this. Maybe in future PR?

**Comment by ChinYikMing** [comment]
> > Note that the run_and_trace function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop rv_userdata, the run_and_trace can only get program name by passing opt_prog_name or attr to it.
> 
> We could streamline the `run` and `run_and_trace` functions by treating `run_and_trace` as a specialized variant of `rv_run`, activated through specific callback functions or flags. This strategy would enable us to eliminate the need for exposing `rv_userdata` directly.

Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.

**Comment by ChinYikMing** [comment]
> Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.

Another way: we can introduce `rv_set_main_loop(rv, void callback(void *arg), void *arg)`. The application can pass their own callback to register main loop. In this way, we can simply call their callback and decouple for many scenario. Inspired by [emscripten](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_loop#c.emscripten_set_main_loop). If callback is NULL, we can fallback to normal `rv_run`. If callback is NULL and `trace` field in `vm_attr_t` is set, then fallback to `rv_run_and_trace`. Specifically, `rv_run` and `rv_run_and_trace` are private function in "riscv.c".

**Comment by jserv** [comment]
> Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.

Exactly. It is time to drop `rv_userdata` by extensive use of `vm_attr_t`. We may refine the term "userdata" as well to avoid unintended confusion.

**Comment by ChinYikMing** [comment]
> > > > Note that the run_and_trace function requires program name to run for now, see [here](https://github.com/sysprog21/rv32emu/pull/340/files#diff-e0cf5b28d9b6b600f0af2bc78e8fd30ec675fd731a5da86f0c4283ffc0e40176). If we drop rv_userdata, the run_and_trace can only get program name by passing opt_prog_name or attr to it.
> > > 
> > > 
> > > We could streamline the `run` and `run_and_trace` functions by treating `run_and_trace` as a specialized variant of `rv_run`, activated through specific callback functions or flags. This strategy would enable us to eliminate the need for exposing `rv_userdata` directly.
> > 
> > 
> > Since we have introduced `vm_attr_t`, we can simply add one more field (e.g., `trace`). `rv_run` can check that easily.
> 
> Another way: we can introduce `rv_set_main_loop(void callback(void *arg), void *arg)`. The application can pass their own callback to register main loop. In this way, we can simply call their callback and decouple for many scenario. Inspired by [emscripten](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_loop#c.emscripten_set_main_loop). If callback is NULL, we can fallback to normal `rv_run`.

How about this way?

**Comment by jserv** [comment]
> Another way: we can introduce `rv_set_main_loop(void callback(void *arg), void *arg)`. The application can pass their own callback to register main loop. In this way, we can simply call their callback and decouple for many scenario. Inspired by [emscripten](https://emscripten.org/docs/api_reference/emscripten.h.html?highlight=main_loop#c.emscripten_set_main_loop). If callback is NULL, we can fallback to normal `rv_run`. If callback is NULL and `trace` field in `vm_attr_t` is set, then fallback to `rv_run_and_trace`. Specifically, `rv_run` and `rv_run_and_trace` are private function in "riscv.c".

Including the callback function as part of the emulator's context is possible, but it’s not necessary to have a separate function like `rv_set_main_loop` for assigning callbacks. Specifying the callback function directly using C99 designated initializers is more efficient, especially since it is unlikely we'll need to change the callback during the execution of the RISC-V emulation main loop.

File: src/riscv.h
Comment (ChinYikMing) [reply src/riscv.h]: Since we drop the `rv_userdata`, no more comment about them.

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: Rename to `log_level` for short.

File: src/riscv_private.h
Comment (jserv) [inline src/riscv_private.h]: Would it be better to name this field as `priv` or `data`?

File: src/riscv.h
Comment (jserv) [inline src/riscv.h]: It is quite confusing to have `vm` and `emu` here.

File: src/main.c
Comment (RinHizakura) [reply src/main.c:226]: Sure. We can leave these at other place.

File: src/riscv.c
Comment (RinHizakura) [inline src/riscv.c]: How about adding another macro like `FILENO()` here for both, so we can make the codes here clean? Then you can define platform-specific `FILENO()`  at the place you defined things like `STDIN_FILENO`.

File: src/main.c
Comment (RinHizakura) [reply src/main.c:226]: Sure. We can leave these at other places.

File: src/riscv.c
Comment (ChinYikMing) [reply src/riscv.c]: Got it.

File: src/riscv_private.h
Comment (ChinYikMing) [reply src/riscv_private.h]: I think `data` is decent.

File: src/riscv.h
Comment (ChinYikMing) [reply src/riscv.h]: How about `vm_data_t`? The instance of it can simply called `data` in `vm_attr_t` struct. Hopefully, it will not confuse with the `data` in internal implementation. I think it is not because the latter is private.

File: src/riscv.c
Comment (ChinYikMing) [reply src/riscv.c]: Check this [commit](https://github.com/sysprog21/rv32emu/pull/340/commits/ea51b2d1300c3d2379c4274af1faa74423fece89#diff-f99b8d1f48ef2257a9bbbe6b487c621f74ef233c00858df5a37b06e89441e836) out.

File: src/main.c
Comment (ChinYikMing) [reply src/main.c:226]: > Sure. We can leave these at other places.

I have mentioned it in the commit message.

File: src/riscv.h
Comment (jserv) [reply src/riscv.h]: > How about `vm_data_t`? 

It looks good to me.

**Review by RinHizakura — APPROVED** [review]
I'm good with the changes now! Let's see if other guys have more comments.

Thank you!

**Review by jserv — CHANGES_REQUESTED** [review]
In your git commit messages, it is more appropriate to mention "this commit" rather than "this PR," as "PR" refers to a pull request, not a commit.

**Review by visitorckw — APPROVED** [review]

**Comment by visitorckw** [comment]
The changes also look good to me. Thanks!

**Pull request merged**
